lib/trace_mgr.js

```diff
@@ -0,0 +1,611 @@
+            sampleRes = emitSyncTraceKind(emitKind, optInfo);
+        }
+        else {
+            sampleRes = emitAsyncTraceKind(emitKind);
+        }
+
+        if (sampleRes.flag === 'success') {
+            updateGlobalSampleStats(emitKind);
+        }
+        return sampleRes;
+
+    } catch (ex) {
+        process.stderr.write('Failed in emitTrace with: ' + ex);
+        return buildTraceResult('fail');
+    } finally {
+        reentrantDisable = false;
+    }
+}
+
+////
+//Helpers for trace calls
+function emitSyncTraceKind(emitKind, optInfo) {
+    //build up trace name 
+    var traceName = emitKind;
+    if (emitKind === 'emitOnExit') {
+        traceName += ('_code-' + optInfo);
+    }
+
+    //invoke the trace writer and remote manager if needed
+    var resolvedPath = createTraceLogTarget(traceName);
+    if (!resolvedPath) {
+        return buildTraceResult('fail');
+    }
+
+    var action = function () {
+        try {
+            process.stderr.write(`    Write error trace to: ${resolvedTracePath}\n`);
+            global.emitTTDLog(resolvedTracePath);
+
+            if(emitOptions.remoteTraceManagerObj && emitOptions.remoteTraceManagerObj.uploadTraceSync) {
+                emitOptions.remoteTraceManagerObj.uploadTraceSync(resolvedPath);
+            }
+        } catch(ex) {
+            process.stderr.write('Failed in emitTrace with: ' + ex);
+        }
+    }
+
+    return buildTraceResult('success', action);
+}
+
+function emitAsyncTraceKind(emitKind) {
+    //get trace stack and check if we want to emit
+    var stk = generateFuzzyStack(emitKind);
+    var entry = checkBinShouldEmit(stk);
+    if (!entry) {
+        return buildTraceResult('no-sample');
+    }
+
+    //build up trace name 
+    var traceBucketName = `${emitKind}_${stk.fbin}_bucket-${entry.bucketId}`;
+    var traceDirName = `trace-${entry.traceCtr}`;
+    var traceName = traceBucketName + path.sep + traceDirName + path.sep;
+
+    //invoke the trace writer and remote manager if needed
+    var resolvedPath = createTraceLogTarget(traceName);
+    if (!resolvedPath) {
+        return buildTraceResult('fail');
+    }
+
+    var action = function () {
+        try {
+            process.stderr.write(`    Write error trace to: ${resolvedTracePath}\n`);
+            global.emitTTDLog(resolvedTracePath);
+
+            if(emitOptions.remoteTraceManagerObj && emitOptions.remoteTraceManagerObj.uploadTraceAsync) {
+                emitOptions.remoteTraceManagerObj.uploadTraceAsync(resolvedPath);
+            }
+        } catch(ex) {
+            process.stderr.write('Failed in emitTrace with: ' + ex);
+        }
+    }
+
+    //update the bin stats as needed
+    updateEmitInfo(entry, resolvedPath);
+
+    return buildTraceResult('success', action);
+}
+
+//create a directory for emitting the trace (if possible) and return it
+function createTraceLogTarget(tracename) {
+    var traceRootDir = emitOptions.localTraceDirectory;
```

Comment:
you could use the same pattern as earlier (traceRootDir = emitOption.localTraceDirectory || blah)